transport.go

```diff
@@ -90,39 +91,40 @@ func (t *transport) IsConnected() bool {
 	}
 }
 
-func (t *transport) Run() error {
-	if !t.IsConnected() {
-		return io.EOF
-	}
-
+// ReceiveFrames starts processing incoming RPC messages
+// synchronously. Always returns a non-nil error, e.g. io.EOF if the
+// connection was closed.
+func (t *transport) ReceiveFrames() error {
 	select {
 	case <-t.startCh:
-		return t.run()
+		// First time calling Run() -- proceed.
 	default:
-		return nil
+		// Second time calling run -- error.
+		return errors.New("Run() called more than once")
 	}
-}
 
-func (t *transport) RunAsync() <-chan error {
-	errCh := make(chan error, 1)
-
-	if !t.IsConnected() {
-		errCh <- io.EOF
-		return errCh
-	}
+	return t.receiveFrames()
+}
 
+// ReceiveFramesAsync starts processing incoming RPC messages in a
+// separate goroutine, if it's not doing so already. This is morally
+// equivalent to go func() { t.RecieveFrames() }(), but avoids
+// spawning a goroutine unless necessary.
+func (t *transport) ReceiveFramesAsync() {
```

Comment:
Ah I just read your comment. Is avoiding the potential goroutine really worth all that code?